### PR TITLE
Sequence parallel prefill attention kernel

### PR DESCRIPTION
**Pull request by ivanium**
This PR mainly implemented `RadixAttention.seq_parallel_extend_forward_flashinfer()` method. We adopted the parallelization strategy discussed before and overlapped communication and computation in each iteration within the ring attention algorithm. Specifically, each SP worker has: 

* q tensor: [batch_size, seq_len, q_head_num // SP_SIZE, head_dim] # partitioned along the `q_head_num` dimension. Assuming that `q_head_num` is divisible by `SP_SIZE`.
* k tensor: [batch_size, seq_len // SP_SIZE, k_head_num, head_dim] # Here `seq_len // SP_SIZE` should be adjusted correspondingly when `seq_len` cannot be divisible by `SP_SIZE`.
* v tensor: [batch_size, seq_len // SP_SIZE, v_head_num, head_dim] # Same as k tensor

*NOTE*: for now the kernel is only tested when `seq_len` can be divided by `SP_SIZE`. Will update the code later.

To balance the workload of all workers, we schedule the computation tasks in the following way:
At iteration `i` (starting from 0), each SP worker will compute the self-attention for its currently active shard (calling ragged attention kernel), and `i` number of cross-shard attention (calling paged attention kernel). SP workers have a perfectly balanced workload in each iteration, although the workload per iteration increases step by step. We will need to further investigate the performance impact of this design.

### CODE REVIEW

File: python/sglang/srt/managers/controller/infer_batch.py [hunk 1]
```diff
@@ -947,28 +947,56 @@ def init_flashinfer_args(
...
+        prefix_lens_cpu = prefix_lens.cpu().numpy()
+        for i in range(batch_size):
+            seq_len = seq_lens_cpu[i]
+            prefix_len = prefix_lens_cpu[i]
+            seq_lens_cpu[i] = seq_len // model_runner.sp_size + (
+                seq_len % model_runner.sp_size > model_runner.sp_rank
+            )
+            prefix_lens_cpu[i] = prefix_len // model_runner.sp_size + (
+                prefix_len % model_runner.sp_size > model_runner.sp_rank
+            )
```
Comment (ZYHowell) [inline python/sglang/srt/managers/controller/infer_batch.py]: now we don't use this layout.
Instead, for all `sp_rank != sp_size - 1`, the local len is `ceil(len / sp_size)`; for `sp_rank == sp_size - 1`, it's the `len % ceil(len / sp_size)`

Comment (ivanium) [reply python/sglang/srt/managers/controller/infer_batch.py]: Got it. Let me merge #3 and get this fixed

File: python/sglang/bench_latency.py [hunk 2]
```diff
@@ -98,8 +98,8 @@ def load_model(server_args, tp_rank, sp_rank: int = 0):
 def prepare_inputs(bench_args, tokenizer):
     prompts = [
         "The capital of France is",
-        "The capital of the United Kindom is",
-        "Today is a sunny day and I like",
+        "The capital of the United Kindom",
+        "Today is a sunny day and I like to",
```
Comment (ZYHowell) [inline python/sglang/bench_latency.py]: this is to avoid padding? can we revert it back now?

File: python/sglang/bench_latency.py [hunk 3]
```diff
@@ -198,12 +198,13 @@ def correctness_test(
...
     # Decode
     output_ids = [input_ids[i] + [next_token_ids[i]] for i in range(len(input_ids))]
-    for _ in range(bench_args.output_len):
-        next_token_ids, _ = decode(next_token_ids, batch, model_runner)
-        for i in range(len(reqs)):
-            output_ids[i].append(next_token_ids[i])
+    # for _ in range(bench_args.output_len):
+    #     next_token_ids, _ = decode(next_token_ids, batch, model_runner)
+    #     for i in range(len(reqs)):
+    #         output_ids[i].append(next_token_ids[i])
```
Comment (ZYHowell) [inline python/sglang/bench_latency.py:207]: would decoder be supported in the next PR or this one?

File: python/sglang/srt/managers/controller/infer_batch.py (hunk 1 above)
Comment (ZYHowell) [reply python/sglang/srt/managers/controller/infer_batch.py]: #3 is merged to main, let's resolve this one

File: python/sglang/srt/layers/parallel_utils/parallel_state.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+def get_sequence_parallel_prev_rank(rank: Optional[int] = None):
+    assert _SEQUENCE_PARALLEL_SIZE is not None
+    if rank is None:
+        rank = get_tensor_model_parallel_rank()
+    first_rank = get_sequence_parallel_first_rank(rank)
+    prev_rank = first_rank + (rank - 1) % _SEQUENCE_PARALLEL_SIZE
+    return prev_rank
+
+
+def get_actual_tensor_model_parallel_world_size():
```
Comment (ZYHowell) [inline python/sglang/srt/layers/parallel_utils/parallel_state.py]: I'd recommend `actual -> kv` but it's just a naming issue

File: python/sglang/srt/layers/radix_attention.py [hunk 5]
```diff
@@ -126,6 +136,175 @@ def decode_forward_flashinfer(self, q, k, v, input_metadata: InputMetadata):
...
+            # (i starting from 0), each SP worker will run i paged attentions.
+            for existing_sid in owned_sids:
+                if existing_sid == sid:
+                    continue
+                # Due to the causal nature of the attention, swap pids if necessary
+                i, j = (
+                    (existing_sid, sid) if existing_sid > sid else (sid, existing_sid)
+                )
+                q_data = qs[i]
+                kv_data = torch.stack(owned_shards[j], dim=1)
```
Comment (ZYHowell) [inline python/sglang/srt/layers/radix_attention.py]: instead of a stack, can we do: 1. allocate kv_data together; 2. construct a view for k_data and v_data? (`k_data = kv_data[:, :k_head, :], v_data = kv_data[:, k_head:, :]`) This reduces a potentially large memcpy

File: python/sglang/srt/managers/controller/infer_batch.py [hunk 6]
```diff
@@ -947,28 +945,56 @@ def init_flashinfer_args(
...
+                for i in range(batch_size)
+            ],
+            dim=0,
+        ).contiguous()
+    else:
+        kv_indptr = torch.zeros((batch_size + 1,), dtype=torch.int32, device="cuda")
+        kv_indptr[1:] = torch.cumsum(seq_lens - prefix_lens, dim=0)
+        kv_indices = torch.arange(
+            0, torch.sum(seq_lens), dtype=torch.int32, device="cuda"
+        )
```
Comment (ZYHowell) [inline python/sglang/srt/managers/controller/infer_batch.py]: this seems pretty hacky and I don't understand why arange is correct...

File: python/sglang/srt/layers/linear.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
+import logging
```
Comment (ZYHowell) [inline python/sglang/srt/layers/linear.py:3]: I'm a bit confused by the purpose of this file. My understanding is that you want to have the "actual tp size" applied for KV Linear Proj while tp size for Q Proj and O Proj. However, the `get_num_kv_head` already handled this case by default( see the max operation in https://github.com/vllm-project/vllm/blob/3eeb148f467e3619e8890b1a5ebe86a173f91bc9/vllm/config.py#L372-L380)

File: python/sglang/srt/managers/controller/infer_batch.py (hunk 1 above)
Comment (ivanium) [reply python/sglang/srt/managers/controller/infer_batch.py]: Thanks! I have a fix for this locally and will push it soon.

File: python/sglang/bench_latency.py (hunk 2 above)
Comment (ivanium) [reply python/sglang/bench_latency.py]: Maybe not now because I haven't added masks for padding tokens. But will make sure to revert it before merging this PR.

File: python/sglang/bench_latency.py (hunk 3 above)
Comment (ivanium) [reply python/sglang/bench_latency.py:207]: How about we leave the decode phase to the next PR?

File: python/sglang/srt/layers/parallel_utils/parallel_state.py (hunk 4 above)
Comment (ivanium) [reply python/sglang/srt/layers/parallel_utils/parallel_state.py]: Make sense. I will fix this.

File: python/sglang/srt/layers/radix_attention.py (hunk 5 above)
Comment (ivanium) [reply python/sglang/srt/layers/radix_attention.py]: I should have used the ragged attention kernel here which avoids creating this redundant `kv_data`. Will push a fix to this soon.

File: python/sglang/srt/managers/controller/infer_batch.py (hunk 6 above)
Comment (ivanium) [reply python/sglang/srt/managers/controller/infer_batch.py]: We will get rid of this magic after switching to use ragged attention for non-causal attention parts.

File: python/sglang/srt/layers/linear.py (hunk 7 above)
Comment (ivanium) [reply python/sglang/srt/layers/linear.py:3]: I think it is the GQA that makes this part complicated. Because now we applied different parallelisms for Q/O proj and KV proj, we need to manually handle the mapping between Q heads and K/V heads. 

For example, suppose that we have TP1 x SP2, 8 Q heads `[q0, q1, ..., q7]`, and 2 KV heads `[k0, k1]`. GQA assumes two groups, where `[q0, q1, q2, q3] -> [k0]`, and `[q4, q5, q6, q7] -> [k1]`. Then for SP worker 0, it should be assigned Q heads `[q0, q1, q4, q5]`, and KV heads `[k0, k1]`. So this file adapts original vllm linear layers but changes the weight loader to select the correct set of Q/O heads (https://github.com/ivanium/sglang/blob/seq-parallel-kernel/python/sglang/srt/layers/linear.py#L486)

Comment (ZYHowell) [reply python/sglang/srt/layers/linear.py:3]: We don't have the case right? The TP is always >= SP

Comment (ivanium) [reply python/sglang/srt/layers/linear.py:3]: My bad. I meant TP2 x SP2

Comment (ZYHowell) [reply python/sglang/srt/layers/linear.py:3]: now kind of understand your point

File: python/sglang/bench_latency.py (hunk 2 above)
Comment (ivanium) [reply python/sglang/bench_latency.py]: Fixed

File: python/sglang/srt/layers/parallel_utils/parallel_state.py (hunk 4 above)
Comment (ivanium) [reply python/sglang/srt/layers/parallel_utils/parallel_state.py]: Fixed.

File: python/sglang/srt/layers/radix_attention.py (hunk 5 above)
Comment (ivanium) [reply python/sglang/srt/layers/radix_attention.py]: Fixed.

File: python/sglang/srt/managers/controller/infer_batch.py (hunk 6 above)
Comment (ivanium) [reply python/sglang/srt/managers/controller/infer_batch.py]: Fixed.

**Review by ZYHowell — APPROVED** [review]
lgtm shall we merge this and do decoding in the next pr?

**Comment by ivanium** [comment]
Sounds good
